packages/contracts-bedrock/test/cannon/MIPS2.t.sol

```diff
@@ -0,0 +1,1787 @@
+        uint32 futexAddr = 0x1000;
+        uint32 futexVal = 0xAA_AA_AA_AA;
+
+        uint32 insn = 0x0000000c; // syscall
+        (MIPS2.State memory state, MIPS2.ThreadState memory thread, bytes memory memProof) =
+            constructMIPSState(0, insn, futexAddr, futexVal);
+        thread.registers[2] = sys.SYS_FUTEX;
+        thread.registers[A0_REG] = futexAddr;
+        thread.registers[A1_REG] = sys.FUTEX_WAIT_PRIVATE;
+        thread.registers[A2_REG] = 0xBB_BB_BB_BB;
+        thread.registers[A3_REG] = 0; // timeout
+        threading.createThread();
+        threading.replaceCurrent(thread);
+        bytes memory threadWitness = threading.witness();
+        finalizeThreadingState(threading, state);
+
+        // FUTEX_WAIT
+        MIPS2.ThreadState memory expectThread = copyThread(thread);
+        expectThread.pc = thread.nextPC;
+        expectThread.nextPC = thread.nextPC + 4;
+        expectThread.futexAddr = futexAddr;
+        expectThread.registers[2] = 0xFF_FF_FF_FF;
+        expectThread.registers[7] = sys.EAGAIN; // errno
+        threading.replaceCurrent(expectThread);
+
+        MIPS2.State memory expect = copyState(state);
+        expect.step = state.step + 1;
+        expect.stepsSinceLastContextSwitch = state.stepsSinceLastContextSwitch + 1;
+        finalizeThreadingState(threading, expect);
+
+        bytes32 postState = mips.step(encodeState(state), bytes.concat(threadWitness, memProof), 0);
+        assertEq(postState, outputState(expect), "unexpected post state");
+    }
+
+    function test_syscallFutexWake_succeeds() public {
+        uint32 futexAddr = 0x1000;
+        uint32 insn = 0x0000000c; // syscall
+        (MIPS2.State memory state, MIPS2.ThreadState memory thread, bytes memory memProof) =
+            constructMIPSState(0, insn, futexAddr, 0xAA_AA_AA_AA);
+        thread.threadID = threading.nextThreadID();
+        thread.registers[2] = sys.SYS_FUTEX;
+        thread.registers[A0_REG] = futexAddr;
+        thread.registers[A1_REG] = sys.FUTEX_WAKE_PRIVATE;
+        thread.registers[A2_REG] = 1000; // ignored
+        thread.registers[7] = 0xbeef; // non-zero value to check if it is cleared
+        threading.createThread();
+        threading.replaceCurrent(thread);
+        bytes memory threadWitness = threading.witness();
+        finalizeThreadingState(threading, state);
+
+        // FUTEX_WAKE
+        threading.left().pop();
+        MIPS2.ThreadState memory expectThread = copyThread(thread);
+        expectThread.pc = thread.nextPC;
+        expectThread.nextPC = thread.nextPC + 4;
+        expectThread.registers[2] = 0x0;
+        expectThread.registers[7] = 0x0; // errno
+        threading.right().push(expectThread);
+
+        MIPS2.State memory expect = copyState(state);
+        expect.wakeup = futexAddr;
+        expect.step = state.step + 1;
+        expect.stepsSinceLastContextSwitch = 0;
+        finalizeThreadingState(threading, expect);
```

Comment:
Indeed. I'm not rigorous on valid states here since the VM doesn't throw exceptions in this case. In light of recent issues found in the old VM, I'll be adding state validation checks during `step`.
Will fix this test too so it's less confusing.

Reply:
Fixed.